src/rpc/methods/eth/filter/mod.rs

```diff
@@ -0,0 +1,263 @@
+    let min_height = match from_block {
+        None | Some("latest") | Some("") => heaviest,
+        Some("earliest") => 0,
+        Some(block) => {
+            if !block.starts_with("0x") {
+                return Err("FromBlock is not a hex".to_string());
+            }
+            hex_str_to_epoch(block).map_err(|_| "invalid epoch".to_string())?
+        }
+    };
+
+    let max_height = match to_block {
+        None | Some("latest") | Some("") => -1,
+        Some("earliest") => 0,
+        Some(block) => {
+            if !block.starts_with("0x") {
+                return Err("ToBlock is not a hex".to_string());
+            }
+            hex_str_to_epoch(block).map_err(|_| "invalid epoch".to_string())?
+        }
+    };
+
+    if min_height == -1 && max_height > 0 {
+        if max_height - heaviest > max_range {
+            return Err(format!(
+                "invalid epoch range: to block is too far in the future (maximum: {})",
+                max_range
+            ));
+        }
+    } else if min_height >= 0 && max_height == -1 {
+        if heaviest - min_height > max_range {
+            return Err(format!(
+                "invalid epoch range: from block is too far in the past (maximum: {})",
+                max_range
+            ));
+        }
+    } else if min_height >= 0 && max_height >= 0 {
+        if min_height > max_height {
+            return Err(format!(
+                "invalid epoch range: to block ({}) must be after from block ({})",
+                max_height, min_height
+            ));
+        } else if max_height - min_height > max_range {
+            return Err(format!(
+                "invalid epoch range: range between to and from blocks is too large (maximum: {})",
+                max_range
+            ));
+        }
+    }
+
+    Ok((min_height, max_height))
+}
+
+fn hex_str_to_epoch(hex_str: &str) -> Result<ChainEpoch, String> {
+    let hex_substring = hex_str
+        .get(2..)
+        .ok_or_else(|| "invalid hex string: unable to parse epoch".to_string())?;
+    i64::from_str_radix(hex_substring, 16).map_err(|e| e.to_string())
+}
+
+fn parse_eth_topics(topics: &EthTopicSpec) -> Result<HashMap<String, Vec<Vec<u8>>>, String> {
+    let mut keys: HashMap<String, Vec<Vec<u8>>> = HashMap::new();
+    for (idx, vals) in topics.0.iter().enumerate() {
+        if vals.0.is_empty() {
+            continue;
+        }
+        let key = format!("t{}", idx + 1);
+        for v in &vals.0 {
+            keys.entry(key.clone()).or_default().push(v.0 .0.to_vec());
+        }
+    }
+    Ok(keys)
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(crate) struct ActorEventBlock {
+    codec: u64,
+    value: Vec<u8>,
+}
+
+const MULTICODEC_RAW: u64 = 0x55;
+
+fn keys_to_keys_with_codec(
+    keys: HashMap<String, Vec<Vec<u8>>>,
+) -> HashMap<String, Vec<ActorEventBlock>> {
+    let mut keys_with_codec: HashMap<String, Vec<ActorEventBlock>> = HashMap::new();
```

Comment:
We already know the number of keys in the collection. We should preallocate required space to avoid reallocations.

Reply:
fixed